### PR TITLE
[WIP] Funding detect double spends

### LINKED ISSUE #1494: walletbalance shows wrong confirmed balance, channels stuck as pending
### Background

`lncli walletbalance` is showing confirmed balance which I don't have. lnd lets me open a channel with those funds, but obviously failing to send out the opening transactions because those outputs are already spent. Channels remain pending for over 1 month now, `lncli walletbalance` is still showing a wrong balance.

### Your environment

* version of `lnd` 
0.4.2
* which operating system (`uname -a` on *Nix)
 4.9.0-6-amd64 #1 SMP Debian 4.9.88-1+deb9u1 (2018-05-07) x86_64 GNU/Linux
* version of `btcd`, `bitcoind`, or other backend
bitcoind (0.16.1)
* any other relevant environment details

### Steps to reproduce

In my case my walletbalance shows
`{
    "total_balance": "139197",
    "confirmed_balance": "139197",
    "unconfirmed_balance": "0"
}
`
Although I only really have about 12000 sat. lnd allows me to open a channel with those non existing funds, those channels show up as pending
```
{
    "total_limbo_balance": "0",
    "pending_open_channels": [
        {
            "channel": {
                "remote_node_pub": "0374ecf61ed6c1208c42339f47decde2bc0c4393ac95f07827b3471e939d7eb961",
                "channel_point": "69a75bfae6642276efc1bbbf3d8860e3710cc6894c0cfaa438b4987418837427:1",
                "capacity": "110000",
                "local_balance": "109819",
                "remote_balance": "0"
            },
            "confirmation_height": 0,
            "commit_fee": "181",
            "commit_weight": "600",
            "fee_per_kw": "250"
        },
        {
            "channel": {
                "remote_node_pub": "0374ecf61ed6c1208c42339f47decde2bc0c4393ac95f07827b3471e939d7eb961",
                "channel_point": "835672640218607009922e594baba1d3fdad67fb181aab44e84852051d5361bf:0",
                "capacity": "116252",
                "local_balance": "115528",
                "remote_balance": "0"
            },
            "confirmation_height": 0,
            "commit_fee": "724",
            "commit_weight": "600",
            "fee_per_kw": "1000"
        }
    ],
    "pending_closing_channels": [
    ],
    "pending_force_closing_channels": [
    ],
    "waiting_close_channels": [
    ]
}
```
but the opening transactions obviously never goes out

log says:
```
2018-05-31 18:51:13.139 [ERR] FNDG: Unable to complete reservation sign complete: Transaction rejected: output already spent
2018-05-31 18:51:13.139 [ERR] FNDG: Failing funding flow: Transaction rejected: output already spent ((*lnwire.Error)(0xc42e7021c0)({
2018-05-31 18:51:13.139 [ERR] RPCS: unable to open channel to NodeKey(0374ecf61ed6c1208c42339f47decde2bc0c4393ac95f07827b3471e939d7eb961): Transaction rejected: output already spent

2018-06-03 23:22:22.708 [ERR] FNDG: Unable to complete reservation sign complete: Transaction rejected: output already spent
2018-06-03 23:22:22.708 [ERR] FNDG: Failing funding flow: Transaction rejected: output already spent ((*lnwire.Error)(0xc4220b07c0)({
2018-06-03 23:22:22.708 [ERR] RPCS: unable to open channel to NodeKey(0374ecf61ed6c1208c42339f47decde2bc0c4393ac95f07827b3471e939d7eb961): Transaction rejected: output already spent
```

The weird thing about it is that the balance disappears as confirmed balance from `lncli walletbalance`, after trying to open a channel with those funds, but after some time (I guess hours) it reappears as confirmed balance while the channel is still pending.

### Expected behaviour

`lncli walletbalance` not showing spent outputs as confirmed

### Actual behaviour

`lncli walletbalance` showing funds which were already spent as confirmed, reappearing after trying to open a channel with those "non existent" funds while the channel stays pending.

When trying to send an onchain transaction
```
lncli sendcoins --addr 34bFuLRhXgD8En6k3pgokDaqHck6yzpUmL --amt 130000 --sat_per_byte 1
[lncli] rpc error: code = Unknown desc = insufficient funds available to construct transaction

```

### LINKED ISSUE #1623: Unable to open channel because "mempool min fee not met", channel still shows as pending.
### Background

Tried to open open a channel but it failed due to "mempool min fee not met". Channel still shows as pending.

### Your environment

* version of `lnd` : commit e0baa4969035fc555b185d21acaa0b09aecdef6f (mainnet)
* which operating system (`uname -a` on *Nix): Linux bitcoin-node 4.9.35-v7+ #1014 SMP Fri Jun 30 14:47:43 BST 2017 armv7l GNU/Linux (raspberry pi)
* version of `btcd`, `bitcoind`: Bitcoin Core Daemon version v0.16.1

* wallet balance before:
```
lncli walletbalance
{
"total_balance": "115918",
"confirmed_balance": "115918",
"unconfirmed_balance": "0"
}
```

* bitcoind settings:
```
rpcuser=bitcoinrpc
rpcpassword=<redacted>
dbcache=100
maxorphantx=10
maxmempool=5
minrelaytxfee=0.00000000
server=1
listen=1
txindex=1
zmqpubrawblock=tcp://127.0.0.1:28332
zmqpubrawtx=tcp://127.0.0.1:28332
```

Note that minrelaytxfee was set to 0 to ensure that low TXs created by LND get forwarded.

### Steps to reproduce

Executed openchannel to LIGHTNING.SCOT node:
```
lncli openchannel --node_key 02a456b816697c6b8407ee2176475d74b790a96143bc30e896e8bf7a8852a8b637 --local_amt 100000 --push_amt 0
```

### Expected behaviour

Channel opens, or fails with reason given.

### Actual behaviour

`lncli` blocked and did not return from the RPC call, had to be killed with CTRL+C.

Channel did not open due to "mempool min fee not met".

Channel stuck in pending state, although TX was not broadcast anywhere.

Relevant log output:
```
2018-07-25 13:03:05.506 [INF] LNWL: Inserting unconfirmed transaction 2ae81716cc429e7ce0441469f2f173068d4a0888f669e4fbb67821100e362175
2018-07-25 13:03:05.601 [ERR] FNDG: Unable to complete reservation sign complete: -26: 66: mempool min fee not met
2018-07-25 13:03:05.602 [DBG] FNDG: Failing funding flow for pendingID=900048e234a9c4653880745b6b86c83523bd9c4709b2f6de8a8dd6b275fc3bd2:
-26: 66: mempool min fee not met
2018-07-25 13:03:05.602 [INF] FNDG: Cancelling funding reservation for node_key=02a456b816697c6b8407ee2176475d74b790a96143bc30e896e8bf7a
8852a8b637, chan_id=900048e234a9c4653880745b6b86c83523bd9c4709b2f6de8a8dd6b275fc3bd2
2018-07-25 13:03:05.602 [ERR] FNDG: unable to cancel reservation: unable to cancel reservation: attempted to cancel non-existent funding
state
2018-07-25 13:03:05.602 [DBG] FNDG: Sending funding error to peer (02a456b816697c6b8407ee2176475d74b790a96143bc30e896e8bf7a8852a8b637):
(*lnwire.Error)(0x13a3bbc0)({
ChanID: (lnwire.ChannelID) (len=32 cap=32) 900048e234a9c4653880745b6b86c83523bd9c4709b2f6de8a8dd6b275fc3bd2,
Data: (lnwire.ErrorData) (len=36 cap=48) {
00000000  66 75 6e 64 69 6e 67 20  66 61 69 6c 65 64 20 64  |funding failed d|
00000010  75 65 20 74 6f 20 69 6e  74 65 72 6e 61 6c 20 65  |ue to internal e|
00000020  72 72 6f 72                                       |rror|
}
})
2018-07-25 13:03:05.603 [DBG] PEER: Sending Error(chan_id=900048e234a9c4653880745b6b86c83523bd9c4709b2f6de8a8dd6b275fc3bd2, err=funding
failed due to internal error) to 206.189.124.227:9735
```

Output from lncli afterwards:
```
lncli pendingchannels
{
"total_limbo_balance": "164076",
"pending_open_channels": [
{
"channel": {
"remote_node_pub": "02a456b816697c6b8407ee2176475d74b790a96143bc30e896e8bf7a8852a8b637",
"channel_point": "2ae81716cc429e7ce0441469f2f173068d4a0888f669e4fbb67821100e362175:1",
"capacity": "100000",
"local_balance": "99638",
"remote_balance": "0"
},
"confirmation_height": 0,
"commit_fee": "362",
"commit_weight": "600",
"fee_per_kw": "500"
}
],
"pending_closing_channels": [
],
"pending_force_closing_channels": [
... removed since it's unrelated ...
],
"waiting_close_channels": [
]
}
```

### PR DESCRIPTION
**Pull request by halseth**
This PR makes the `fundingmanager` aware of transaction spending inputs  to the funding transaction, invalidating the funding tx. In this case we will cancel the funding flow.

TODO:
- [ ] handle this also after startup by extracting common logic
- [ ] properly set `pkScript` for funding tx inputs
- [ ] add scenario to integration tests

Fixes #1494
Fixes #1623

~~Blocked by #1910~~

### CODE REVIEW

**Comment by halseth** [comment]
Not ready for review.

File: fundingmanager.go [hunk 1]
```diff
@@ -1628,16 +1645,80 @@ func (f *fundingManager) handleFundingSigned(fmsg *fundingSignedMsg) {
...
+		}
+
 		confChan := make(chan *lnwire.ShortChannelID)
 		cancelChan := make(chan struct{})
 
-		// In case the fundingManager is stopped at some point during
-		// the remaining part of the opening process, we must wait for
-		// this process to finish (either successfully or with some
-		// error), before the fundingManager can be shut down.
+		// We now launch a goroutine that will wiat for the funding
```
Comment (alexbosworth) [inline fundingmanager.go:1720]:
```suggestion
		// We now launch a goroutine that will wait for the funding
```

File: fundingmanager_test.go [hunk 2]
```diff
@@ -2363,3 +2436,106 @@ func TestFundingManagerMaxPendingChannels(t *testing.T) {
 		t, bob.msgChan, "AcceptChannel",
 	).(*lnwire.AcceptChannel)
 }
+
+// TestFundingManagerFundingDoubleSpend tests that we correctly handles the
```
Comment (alexbosworth) [inline fundingmanager_test.go:2440]:
```suggestion
// TestFundingManagerFundingDoubleSpend tests that we correctly handle the
```

**Comment by halseth** [comment]
Gonna close this as it is outdated and unfinished. I still think this is the way to go to solve #386 though.

**Pull request closed without merging**